plugins/cros-ec/fu-cros-ec-usb-device.c

```diff
@@ -0,0 +1,343 @@
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "wMaxPacketSize isn't valid: %" G_GUINT16_FORMAT,
+			     self->chunk_len);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_do_xfer (FuCrosEcUsbDevice * self, guint8 *outbuf,
+			       gsize outlen, guint8 *inbuf, gsize inlen,
+			       gboolean allow_less, gsize *rxed_count,
+			       GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+	gsize actual = 0;
+
+	/* send data out */
+	if (outbuf != NULL && outlen > 0) {
+		if (!g_usb_device_bulk_transfer (usb_device, self->ep_num,
+						 outbuf, outlen,
+						 &actual, BULK_SEND_TIMEOUT_MS,
+						 NULL, error)) {
+			return FALSE;
+		}
+		if (actual != outlen) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only sent %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	/* read reply back */
+	if (inbuf != NULL && inlen > 0) {
+		actual = 0;
+		if (!g_usb_device_bulk_transfer (usb_device,
+						 self->ep_num | 0x80,
+						 inbuf, inlen,
+						 &actual, BULK_RECV_TIMEOUT_MS,
+						 NULL, error)) {
+			return FALSE;
+		}
+		if (actual != inlen && !allow_less) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only received %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	if (rxed_count != NULL)
+		*rxed_count = actual;
+
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_flush (FuDevice *device, gpointer user_data,
+			     GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	gsize actual = 0;
+	g_autofree guint8 *inbuf = g_malloc0 (self->chunk_len);
+
+	if (g_usb_device_bulk_transfer (usb_device, self->ep_num | 0x80, inbuf,
+					self->chunk_len, &actual,
+					FLUSH_TIMEOUT_MS, NULL, NULL)) {
+		g_debug ("flushing %" G_GSIZE_FORMAT " bytes", actual);
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "flushing %" G_GSIZE_FORMAT " bytes\n", actual);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_setup (FuDevice *device, GError **error)
+{
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	gsize rxed_size;
+	guint32 error_code;
+	struct update_frame_header ufh;
+
+	/*
+	 * Need to be backwards compatible, communicate with targets running
+	 * different protocol versions.
+	 */
+	union {
+		struct first_response_pdu rpdu;
+		guint32 legacy_resp;
+	} start_resp;
+
+	/* send start request */
+
+	/* flush all data from endpoint to recover in case of error */
+	if (!fu_device_retry (device, fu_cros_ec_usb_device_flush,
+			      FLUSH_RETRY_CNT, NULL, error)) {
+		g_prefix_error (error, "Failed to flush device to idle state");
```

Comment:
since this is a prefix error you should put a ": " at the end so it reads out properly.

Reply:
also, no need to set capital F, it might not be the very first error prefixed so it might look funny if you get

`error calibrating flux capacitor: Failed to flush device to idle stateflushing 5 bytes`

Reply:
Thanks. Did this and cleaned this up elsewhere too.